File: leetCode/fukun/001-100/lc_033.js
const search = function (nums, target) {
    let left = 0,
        right = nums.length - 1;
    while (left < right) {
        let mid = Math.floor((left + right) / 2);
        (nums[0] > nums[mid]) ^ (target > nums[mid]) ^ (nums[0] > target) ? left = mid + 1 : right = mid;
    }
    return nums[left] === target ? left : -1;
};
console.log(search([4, 5, 6, 7, 0, 1, 2, 3], 0));

File: leetCode/XSS/lc_033.java
public class lc_033 {
    public static void main(String[] args) {
        System.out.println(new Solution().search(new int[]{4,5,6,7,0,1,2},0));
    }
    static class Solution {
        public int search(int[] nums, int target) {
            int n=nums.length;
            if(n==0)return -1;
            int left=0,right=n-1;
            while (left<=right)
            {
                int mid=left + (right-left)/2;
                System.out.println(String.valueOf(right-left)+"/2="+String.valueOf((right-left)>>1));
                if(nums[mid]==target)return mid;
                if(nums[left]<=nums[mid])
                {
                    //前半部分有序
                    if(nums[mid]>target&&nums[left]<=target)
                    {
                        right=mid-1;
                    }
                    else
                    {
                        left=mid+1;
                    }
                }
                else
                {
                    //后半部分有序
                    if(nums[mid]<target&&nums[right]>=target)
                    {
                        left=mid+1;
                    }
                    else
                    {
                        right=mid-1;
                    }
                }
            }
            return -1;
        }
    }
}
